src/CSharp/CodeCracker/Refactoring/ChangeAsToCastCodeFixProvider.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System.Collections.Immutable;
+using System.Composition;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+
+namespace CodeCracker.CSharp.Refactoring
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(ChangeAsToCastCodeFixProvider)), Shared]
+    public class ChangeAsToCastCodeFixProvider : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(DiagnosticId.ChangeAsToCast.ToDiagnosticId());
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(diagnostic.Location.SourceSpan);
+            if (node.Kind() == SyntaxKind.AsExpression)
+                context.RegisterCodeFix(CodeAction.Create("Change safe cast ('as') to direct cast", ct => ChangeAsToCastAsync(context.Document, root, node)), diagnostic);
+            else if (node.Kind() == SyntaxKind.CastExpression)
+                context.RegisterCodeFix(CodeAction.Create("Change direct cast to safe cast ('as')", ct => ChangeCastToAsAsync(context.Document, root, node)), diagnostic);
```

Comment:
Instead of doing work on the `RegisterCodeFixesAsync` use properties to pass info from the analyzer to the code fix. See the `RemoveWhereWhenItIsPossibleAnalyzer` (lines 58 and 59) and `RemoveWhereWhenItIsPossibleCodeFixProvider` (line 25).
The idea is that `RegisterCodeFixesAsync` runs as fast as possible.